### PR TITLE
Fix incorrect handling of _residueWeight in IProfiler data structures

### PR DESCRIPTION
**Pull request by harryyu1994**
- Correct the update and overflow check of `_residueWeight`

Fixes: #450
Signed-off-by: Harry Yu <harryyu1994@gmail.com>

### CODE REVIEW

**Comment by harryyu1994** [comment]
@mpirvu Hi Marius, this change is ready for review.

File: runtime/compiler/trj9/runtime/IProfiler.cpp [hunk 1]
```diff
@@ -2789,15 +2798,7 @@ TR_IPBCDataCallGraph::setData(uintptrj_t v, uint32_t freq)
             }
          }
       else
```
Comment (mpirvu) [inline runtime/compiler/trj9/runtime/IProfiler.cpp]: Please put curly brackets even if it's just one line. Our rules say that if the 'then' part has curly brackets, the 'else' part must have as well for symmetry.

**Comment by mpirvu** [comment]
I now realize that if one the classes exceeds 0x7fff, then the residueWeight will be irrelevant because, due to fact that we maintain it on 15 bits, it cannot be greater than the weight for the dominant class. 
We should try to get rid of that bit field.

**Comment by mpirvu** [comment]
An IProfiler entry of type TR_IPBytecodeHashTableEntry has these fields
```
protected:
   TR_IPBytecodeHashTableEntry *_next;
   uintptrj_t _pc;
   int32_t    _lastSeenClassUnloadID;

   enum TR_PersistenceFlags
      {
      IPBC_ENTRY_CAN_PERSIST_FLAG = 0x01,
      IPBC_ENTRY_PERSIST_LOCK_FLAG = 0x02,
      };

   uint8_t _entryFlags;
   uint8_t _persistFlags;
   }; // class TR_IPBytecodeHashTableEntry
```
Due to compiler padding we could add another couple of bytes and still not see a footprint regression. We could move the _tooBigToBeInlined bit here, though from a conceptual point of view the place of tooBigToBeInlined  is in class CallSiteProfileInfo.

**Comment by harryyu1994** [comment]
Confirmed that adding an uint8_t variable in `TR_IPBytecodeHashTableEntry` won't change the size of it. Will be moving the `_tooBigToBeInlined` bit here.

**Comment by harryyu1994** [comment]
```
typedef struct TR_IPBCDataCallGraphStorage
   {
   TR_IPBCDataStorageHeader header;
   CallSiteProfileInfo _csInfo;
   } TR_IPBCDataCallGraphStorage;
```

```
void
TR_IPBCDataCallGraph::createPersistentCopy(uintptrj_t cacheStartAddress, TR_IPBCDataStorageHeader *storage, TR::PersistentInfo *info)
   {
   TR_IPBCDataCallGraphStorage * store = (TR_IPBCDataCallGraphStorage *) storage;
   ...
   store->_csInfo._residueWeight = _csInfo._residueWeight;
   store->_tooBigToBeInlined = _tooBigToBeInlined;
   }
```

The problem here is that `TR_IPBCDataStorageHeader` does not have `_tooBigToBeInlined`, and adding it there might increase footprint.

**Comment by harryyu1994** [comment]
```
149 typedef struct TR_IPBCDataStorageHeader
150    {
151    uint32_t pc;
152    uint32_t left:8;
153    uint32_t right:16;
154    uint32_t ID:8;
155    } TR_IPBCDataStorageHeader;
```
Should we borrow 1 bit from TR_IPBCDataStorageHeader by making `uint32_t ID:7` then add a bit in `TR_IPBCDataCallGraphStorage` for `_tooBigToBeInlined`? or is that too messy..

**Comment by mpirvu** [comment]
Hmm, I didn't know we have that `struct TR_IPBCDataCallGraphStorage` that is so compact we don't have room for a bool.
That one is used for storing the info into the shared class cache. It's too messy to have the same information in one place if the data is in SCC and in another place if it's in memory.
I am inclined to drop this "enahacement"

**Comment by harryyu1994** [comment]
So should we just reset the classes when _residueWeight reaches 0x7fff? or limit _weight to be 0x7fff max?

**Comment by mpirvu** [comment]
I see that the information for _tooBigToBeInlined is not used anywhere in the optimizer. It used to be used in Inliner, but not anymore. I am waiting for an answer from Andrew, but it is unused we should delete it.

**Comment by mpirvu** [comment]
My grep was bad, so we do actually use  _tooBigToBeInlined. At this point I would leave the bit alone. Let's wrap this PR and then we'll experiment with shifting right once one of the counters reaches maximum.

**Comment by harryyu1994** [comment]
Closing this PR because a new one is created (PR #985).

**Comment by harryyu1994** [comment]
@mpirvu Hi Marius, this change is ready for review.

**Comment by harryyu1994** [comment]
Changes updated

**Comment by mpirvu** [comment]
jenkins test sanity

**Comment by mpirvu** [comment]
@harryyu1994 Could you please run a personal build with these changes.

**Comment by harryyu1994** [comment]
Okay will do

File: runtime/compiler/trj9/runtime/IProfiler.cpp [hunk 2]
```diff
@@ -2770,13 +2770,22 @@ TR_IPBCDataCallGraph::setData(uintptrj_t v, uint32_t freq)
...
+      _csInfo._residueWeight = newResidueWeight;
+      returnCount = newResidueWeight;
+
       if (_csInfo._residueWeight > maxWeight)
          {
          if (lockEntry())
             {
-            for (int32_t i = 1; i < NUM_CS_SLOTS; i++)
+            // Reset the entries in reverse order
+            for (int32_t i = NUM_CS_SLOTS - 1; i > 0; i--)
```
Comment (mstoodle) [inline runtime/compiler/trj9/runtime/IProfiler.cpp]: just curious: why was this change needed?

**Comment by mstoodle** [comment]
Also curious why this PR is still listed as WIP. What remaining work is there (other than to request testing on all platforms) ?

**Comment by mstoodle** [comment]
Jenkins test sanity

File: runtime/compiler/trj9/runtime/IProfiler.cpp (hunk 2 above)
Comment (mstoodle) [reply runtime/compiler/trj9/runtime/IProfiler.cpp]: @mpirvu explained to me that it is to try to minimize the chance of a data race with the surrounding code that identifies empty entries looking for zeros. The risk of a race is more of a "we'll misunderstand the profile" than a "we'll crash" risk, but we should still try to minimize its likelihood.

Comment (mpirvu) [reply runtime/compiler/trj9/runtime/IProfiler.cpp]: Here is an explanation for the reverse order: https://github.com/eclipse/openj9/pull/946#pullrequestreview-90889518

Comment (DanHeidinga) [reply runtime/compiler/trj9/runtime/IProfiler.cpp]: Should that explanation be added as a comment on this code?  It'll make it obvious to the next person to read the code...

Comment (harryyu1994) [reply runtime/compiler/trj9/runtime/IProfiler.cpp]: I will add in the comment

**Comment by harryyu1994** [comment]
The sanity tests kept failing and they aren't related to this change, that's why it's been here for a while.

**Comment by fjeremic** [comment]
@harryyu1994 the tests are failing because of a breaking OMR dependent change that went in as part of https://github.com/eclipse/openj9/pull/1042. Please rebase your branch on the latest master and we will restart the builds. Thanks!

**Comment by harryyu1994** [comment]
Rebased the branch to latest master.

**Comment by mpirvu** [comment]
Jenkins test sanity

**Comment by mpirvu** [comment]
All tests passed. Merging this PR.
